realm/realm-library/src/main/cpp/io_realm_internal_OsRealmConfig.cpp

```diff
@@ -0,0 +1,332 @@
+
+        // error handler will be called form the sync client thread
+        auto error_handler = [](std::shared_ptr<SyncSession> session, SyncError error) {
+            realm::jni_util::Log::d("error_handler lambda invoked");
+
+            auto error_message = error.message;
+            auto error_code = error.error_code.value();
+            if (error.is_client_reset_requested()) {
+                // Hack the error message to send information about the location of the backup.
+                // If more uses of the user_info map surfaces. Refactor this to send the full
+                // map instead.
+                error_message = error.user_info[SyncError::c_recovery_file_path_key];
+                error_code = 7; // See ErrorCode.java
+            }
+
+            JNIEnv* env = realm::jni_util::JniUtils::get_env(true);
+            env->CallStaticVoidMethod(sync_manager_class, java_error_callback_method, error_code,
+                                      to_jstring(env, error_message), to_jstring(env, session.get()->path()));
+        };
+
+        // path on disk of the Realm file.
+        // the sync configuration object.
+        // the session which should be bound.
+        auto bind_handler = [](const std::string& path, const SyncConfig& syncConfig,
+                               std::shared_ptr<SyncSession> session) {
+            realm::jni_util::Log::d("Callback to Java requesting token for path");
+
+            JNIEnv* env = realm::jni_util::JniUtils::get_env(true);
+
+            jstring access_token_string = (jstring)env->CallStaticObjectMethod(
+                sync_manager_class, java_bind_session_method, to_jstring(env, path.c_str()));
+            if (access_token_string) {
+                // reusing cached valid token
+                JStringAccessor access_token(env, access_token_string);
+                session->refresh_access_token(access_token, realm::util::Optional<std::string>(syncConfig.realm_url));
+            }
+        };
+
+        // Get logged in user
+        JStringAccessor user_id(env, j_user_id);
+        JStringAccessor auth_url(env, j_auth_url);
+        SyncUserIdentifier sync_user_identifier = {user_id, auth_url};
+        std::shared_ptr<SyncUser> user = SyncManager::shared().get_existing_logged_in_user(sync_user_identifier);
+        if (!user) {
+            JStringAccessor realm_auth_url(env, j_auth_url);
+            JStringAccessor refresh_token(env, j_reresh_token);
+            user = SyncManager::shared().get_user(sync_user_identifier, refresh_token);
+        }
+
+        util::Optional<std::array<char, 64>> sync_encryption_key(util::none);
+        if (!config.encryption_key.empty()) {
+            sync_encryption_key = std::array<char, 64>();
+            std::copy_n(config.encryption_key.begin(), 64, sync_encryption_key->begin());
+        }
+
+        // Disable ssl by default
```

Comment:
Yes, you are right. I was thinking `sync_client_validate_ssl` is redundant since we have `ssl_trust_certificate_path` to show if we are using ssl or not. But that is wrong, we could support ssl without a customized ssl cert path. Will change this.